android/src/main/java/com/vydia/RNUploader/UploadWorker.kt

```diff
@@ -61,59 +72,76 @@ class UploadWorker(private val context: Context, params: WorkerParameters) :
     val httpMethod: HttpMethod
-    val retries: Int
-    var lastProgressReport = 0L
-    var semaphoreAcquired = false
     try {
       httpMethod = HttpMethod.parse(upload.method)
-      retries = UploadRetry.get(context, upload.id)
       // `setForeground` is recommended for long-running workers.
       // Foreground mode helps prioritize the worker, reducing the risk
       // of it being killed during low memory or Doze/App Standby situations.
-      // ⚠️ This throws error if called in the background,
-      // but if the worker already calls this at foreground,
-      // then subsequent retries will not break it
+      // ⚠️ This should be called in the foreground
       setForeground(getForegroundInfo())
     } catch (error: Throwable) {
-      return@withContext handleError(error, retry = false)
+      handleError(error, retry = false)
+      return@withContext Result.failure()
     }
 
 
-    // complex work, errors thrown below here can trigger retry
-    try {
-      val file = File(upload.path)
-      val size = file.length()
-      val body = file.readChannel()
-
-      // this should happen before the semaphore queueing
-      // to register with the total progress asap
-      handleProgress(0, size)
-
-      semaphore.acquire()
-      semaphoreAcquired = true
-
-      // Use ktor instead of okhttp. Ktor request is coroutine friendly and
-      // will get cancelled when the coroutine gets cancelled
-      val response = client.request(upload.url) {
-        method = httpMethod
-        setBody(body)
-        upload.headers.forEach { (key, value) -> headers.append(key, value) }
-        onUpload { bytesSentTotal, _ ->
-          // throttle progress report
-          val now = System.currentTimeMillis()
-          if (now - lastProgressReport >= PROGRESS_INTERVAL) {
-            lastProgressReport = now
-            handleProgress(bytesSentTotal, size)
+    // Complex work, errors thrown below here trigger retry.
+    // We don't let WorkManager manage retries as it's very buggy.
+    // i.e. we'd occasionally get BackgroundServiceStartNotAllowedException,
+    // or ForegroundServiceStartNotAllowedException, or "isStopped" gets set to "true"
+    // for no reason
+    while (true) {
+      try {
+        // Delay if it's an actual retry, not the first run.
+        // Should be within the try block to account for worker cancellation,
+        // which resumes immediately and throws CancellationException.
+        if (!firstRun) delay(RETRY_DELAY)
+        firstRun = false
+
+        val file = File(upload.path)
+        val size = file.length()
+
+        // Register progress asap so the total progress is accurate
+        // This needs to happen before the semaphore wait
+        handleProgress(0, size)
+
+        // Don't bother to run on an invalid network
+        if (!validateConnection(context, upload.wifiOnly)) continue
+
+        // wait for its turn to run
+        semaphore.acquire()
+        semaphoreAcquired = true
```

Comment:
Maybe an additional inner try/catch wrap around `semaphore.acquire()` (and the code thereafter) would make this logic more clear, and eliminate the need for `semaphoreAcquired`? Something like:

```kotlin
try {
  semaphore.acquire()
  var lastProgressReport = 0L
  val body = file.readChannel()
  val response = client.request(upload.url) {
  // ...
  }
} catch (Throwable t) { 
  // ...
} finally {
  semaphore.release()
}
```

If that makes sense, then maybe this block (the acquire to the release) could be moved to its own function?

Reply:
we might not be able to get rid of the `semaphoreAcquired` because `.acquire()` itself may throw an error, but we can certainly this try/catch down into smaller ones

Reply:
Ah sorry, I meant to have the `acquire()` before/outside of the inner try/catch!

Reply:
ah interesting, yeah that might work. If `acquire` errors out it also stops the `acquire` from getting called